File: 0112-path-sum/0112-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        def path_sum(root,current_sum):
            if not root:
                return False
            current_sum+=root.val
            if not root.left and not root.right:
                return current_sum==targetSum
            return path_sum(root.left,current_sum) or path_sum(root.right,current_sum)
        return path_sum(root,0)
        

File: 0112-path-sum/0112-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    boolean dfs(TreeNode root,int currentSum,int targetSum){
        if(root==null){
            return false;
        }
        currentSum+=root.val;
        if (root.left==null && root.right==null){
            return currentSum==targetSum;
        }
        return dfs(root.left,currentSum,targetSum) || dfs(root.right,currentSum,targetSum);
    }
    public boolean hasPathSum(TreeNode root, int targetSum) {
        return dfs(root,0,targetSum);
    }
}